pallene/to_ir.lua

```diff
@@ -87,52 +90,115 @@ function ToIR:convert_stat(cmds, stat)
             ir.Cmd.Seq(body)))
 
     elseif tag == "ast.Stat.Assign" then
-        local var = stat.var
-        local exp = stat.exp
-        if     var._tag == "ast.Var.Name" then
-            local cname = var._name
-            if     cname._tag == "checker.Name.Local" then
-                self:exp_to_assignment(cmds, cname.id, exp)
-            elseif cname._tag == "checker.Name.Global" then
-                local v = self:exp_to_value(cmds, exp)
-                table.insert(cmds, ir.Cmd.SetGlobal(stat.loc, cname.id, v))
+        local lvars = { locals = {}, globals = {}, brackets = {}, dots = {} }
```

Comment:
Unfortunately, I think this code still has a few bugs. Firstly, there is the big problem of using `pairs` instead of `ipairs`, which means that this algorithm is non-deterministic. Secondly, there are some situations where it should be saving things in temporary variables but it isn't (which can be hard to reproduce due to the pairs nondeterminism). Finally, it also generates temporary variables where it isn't necessary, which boats the generated code.

For an example of the first two problems, consider the following example:

```
function f(a: {integer}, b:{integer})
    a, a[1] = b, 42
end
```

Depending on how the `pairs` orders things, this can results in the following IR, which does the wrong thing. (It assigns to the `b` array while it should be assigning to the `a` array). To reproduce, run `./pallenec --emit-c` multiple times until the `pairs` orders things the buggy and then compile it with `./pallenec --compile-c`.

```
tmp1 = b
tmp2 = 1
tmp3 = 30
a = tmp1
a[tmp2] = tmp3
```

The other more minor problem is that the new version of the code always generates temporary variables, even when not strictly necessary. For example, a simple assignment like `a = b` becomes

```
tmp1 = b
a = tmp1
```

------------------------------------------------------

But I have good  news! I have investigated how Lua implements multiple assignments and figured out what we need to do to generate code that is both correct and without unnecessary temporary variables. I also implemented and uploaded the code that does that to the "hugo-multiple" branch. Check it out! I figured that it would be clearer to just write down the code instead of giving detailed explanation of the algorithm. Feel free to use that code in the PR but if you do please look carefully through it as if you were the one that were doing the code review this time. :)

One thing that I did not do is add extra tests. I would appreciate if you could write some additional tests to cover all the places where we call the `save_if_necessary` function. For example, we would need tests for `a, a[i] = ` and `a[i], a = `. 

We might also want to add a test for `x, x = 10, 20` to verify that it returns 10 just like Lua does.